src/librustc/middle/traits/error_reporting.rs

```diff
@@ -230,6 +230,26 @@ pub fn report_selection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
                                 "the trait `{}` is not implemented for the type `{}`",
                                 trait_ref, trait_ref.self_ty());
 
+                            let mut counter = 1;
+                            infcx.tcx.sess.fileline_help(
+                                obligation.cause.span,
+                                "the following implementations were found:");
+                            infcx.tcx.lookup_trait_def(trait_ref.def_id()).for_each_relevant_impl(
+                                infcx.tcx,
+                                trait_ref.self_ty(),
+                                |impl_def_id| {
+                                    match infcx.tcx.impl_trait_ref(impl_def_id) {
+                                        Some(ref imp) => {
+                                            infcx.tcx.sess.fileline_help(
+                                                obligation.cause.span,
+                                                &format!("implementation {}: `{}`", counter, imp));
```

Comment:
I'm not sure how useful the number is here. I think just indenting the name slightly should work well.

Also, I'd like to limit the number of output traits to some small-ish number, say three or four. This is because sometimes there might be *a lot* of "applicable" traits, say implementations of `From`, for example. In one of my libraries, I have an implementation of `From<T>` for T = all integer types. If you tried to use, say, a float by accident, that would print out 8 implementations. Use of macros can easily produce a large number of potential implementations.

I'd like an output like this:

```
the following implementations were found:
  Foo<A>,
  Foo<B>,
  Foo<C>,
  Foo<D>
and 4 others
```